### PR TITLE
Improve build controller

### PR DESCRIPTION
**Pull request by fejta**
* Better debuggability
* Observe config.yaml prowjob_namespace
* Support using --build-cluster to create contexts, loading gcp auth in kubecfg
* Handle cluster contexts that do not have the build crd deployed

Also:
* Update knative/build to v0.2.0
* Add knative-build periodic job

/assign @stevekuznetsov @cjwagner

### CODE REVIEW

**Comment by fejta** [comment]
This is actually +250 -150, plus a bunch of vendor nonsense.

**Comment by gyuho** [comment]
> git://bitbucket.org/ww/goautoneg: fatal:

Had the same issue in `pull-test-infra-verify-deps` :0

**Comment by BenTheElder** [comment]
see #sig-testing, we've had known issues with autoneg on bitbucket, @munnerz forked it to avoid this IIRC

**Comment by fejta** [comment]
Appears https://github.com/kubernetes/test-infra/pull/10361 should have solved

File: prow/cmd/build/controller.go [hunk 1]
```diff
@@ -206,30 +245,39 @@ func fromKey(key string) (string, string, string, error) {
 func (c *controller) enqueueKey(ctx string, obj interface{}) {
 	var meta metav1.ObjectMeta
 	switch o := obj.(type) {
+	case *prowjobv1.ProwJob:
+		meta = o.ObjectMeta
 	case prowjobv1.ProwJob:
 		meta = o.ObjectMeta
```
Comment (krzyzacy) [inline prow/cmd/build/controller.go]: why keep both? (those are skipped earlier)

Comment (fejta) [reply prow/cmd/build/controller.go]: Done

File: prow/cmd/build/controller.go [hunk 2]
```diff
@@ -50,41 +50,74 @@ import (
...
+			if ns == "" {
+				ns = "controller's"
+			}
+			logrus.Infof("Waiting on prowjobs in %s namespace...", ns)
+		}
 		return false // still syncing prowjobs
 	}
-	for _, cfg := range c.builds {
+	if !c.prowJobsDone {
+		c.prowJobsDone = true
```
Comment (krzyzacy) [inline prow/cmd/build/controller.go]: this boolean seems not necessary

File: prow/cmd/build/controller.go [hunk 3]
```diff
@@ -50,41 +50,74 @@ import (
...
+			c.buildsDone[n] = true
+			logrus.Infof("Synced %s builds", n)
 		}
 	}
 	return true // Everyone is synced
 }
 
-func newController(kc kubernetes.Interface, pjc prowjobset.Interface, pji prowjobinfov1.ProwJobInformer, buildConfigs map[string]buildConfig, totURL string) *controller {
+func limiter() workqueue.RateLimiter {
+	// return workqueue.DefaultControllerRateLimiter
```
Comment (krzyzacy) [inline prow/cmd/build/controller.go]: remove?

File: prow/cmd/build/controller.go [hunk 4]
```diff
@@ -94,37 +127,38 @@ func newController(kc kubernetes.Interface, pjc prowjobset.Interface, pji prowjo
...
-		pjLister:   pji.Lister(),
-		pjInformer: pji.Informer(),
-		workqueue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
-		recorder:   recorder,
-		totURL:     totURL,
+		pjc:         pjc,
+		builds:      buildConfigs,
+		pjLister:    pji.Lister(),
+		pjInformer:  pji.Informer(),
+		workqueue:   workqueue.NewNamedRateLimitingQueue(limiter(), controllerName),
```
Comment (krzyzacy) [inline prow/cmd/build/controller.go]: feels like this should be passed in from main, like other fields.

File: prow/cmd/build/controller.go [hunk 5]
```diff
@@ -50,41 +50,74 @@ import (
...
 )
 
 type controller struct {
-	pjc    prowjobset.Interface
-	builds map[string]buildConfig
-	totURL string
+	pjNamespace string
+	pjc         prowjobset.Interface
+	builds      map[string]buildConfig
+	totURL      string
```
Comment (krzyzacy) [inline prow/cmd/build/controller.go]: oh really? :disappointed:

File: prow/cmd/build/controller.go [hunk 6]
```diff
@@ -285,7 +326,7 @@ func reconcile(c reconciler, key string) error {
 	var wantBuild bool
 
 	// TODO(fejta): only consider prowjob namespace
```
Comment (krzyzacy) [inline prow/cmd/build/controller.go]: the TODO is resolved now?

Comment (krzyzacy) [reply prow/cmd/build/controller.go]: also qq on 
```
case pj.DeletionTimestamp == nil:
 		wantBuild = true
 	}
```

who handles this | sinker? another controller?

File: prow/cmd/build/controller.go [hunk 7]
```diff
@@ -241,23 +284,21 @@ func (c *controller) getBuild(context, namespace, name string) (*buildv1alpha1.B
...
 func (c *controller) deleteBuild(context, namespace, name string) error {
+	logrus.Debugf("deleteBuild(%s,%s,%s)", context, namespace, name)
 	b, ok := c.builds[context]
 	if !ok {
 		return errors.New("context not found")
 	}
 	return b.client.BuildV1alpha1().Builds(namespace).Delete(name, &metav1.DeleteOptions{})
 }
 func (c *controller) createBuild(context, namespace string, b *buildv1alpha1.Build) (*buildv1alpha1.Build, error) {
+	logrus.Debugf("createBuild(%s,%s,%s)", context, namespace, b.Name)
```
Comment (stevekuznetsov) [inline prow/cmd/build/controller.go]: Does the build client not have logging?

File: prow/cmd/build/main.go [hunk 8]
```diff
@@ -158,12 +231,22 @@ func main() {
 	o := parseOptions()
 	logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "build"})
 
-	configs, defaultContext, err := contextConfigs(o.kubeconfig)
+	pjNamespace := ""
+	if o.config != "" {
+		pc, err := config.Load(o.config, "") // ignore jobConfig
```
Comment (stevekuznetsov) [inline prow/cmd/build/main.go]: Why not the agent?

File: prow/cmd/build/controller.go [hunk 9]
```diff
@@ -22,21 +22,21 @@ import (
 	"strings"
 	"time"
 
-	"k8s.io/test-infra/prow/kube"
-	"k8s.io/test-infra/prow/pjutil"
-	"k8s.io/test-infra/prow/pod-utils/clone"
-	"k8s.io/test-infra/prow/pod-utils/decorate"
-	"k8s.io/test-infra/prow/pod-utils/downwardapi"
-
```
Comment (mithrav) [inline prow/cmd/build/controller.go]: Why change import ordering?

File: prow/cmd/build/controller.go (hunk 4 above)
Comment (fejta) [reply prow/cmd/build/controller.go]: Done

File: prow/cmd/build/main.go (hunk 8 above)
Comment (stevekuznetsov) [reply prow/cmd/build/main.go]: I worry downstream users will use other parts of the config later, other than the namespace, and not realize it's not reloading.

Comment (fejta) [reply prow/cmd/build/main.go]: How? The scope of the config object is this if statement.

**Comment by stevekuznetsov** [comment]
Dreaded question -- do we want to use the higher-level controller builder libraries? Would address some of the frustrating and hard-to-catch errors we've all had writing controllers

File: prow/cmd/build/main.go [hunk 10]
```diff
@@ -172,21 +264,26 @@ func main() {
```
Comment (krzyzacy) [inline prow/cmd/build/main.go]: babynit: indent? (for the line above ^^)

**Pull request merged**